### PR TITLE
feat(gpu): improve full propagation in sum and sub

### PR DESCRIPTION
**Pull request by guillermo-oyarzun**
closes: _please link all relevant issues_

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Review by agnesLeroy — COMMENTED** [review]
Hey @guillermo-oyarzun! Thanks a lot for this PR :pray: Here comes a first review: I don't know the details of the implementation so it's hard for me to go through all the logic though. Maybe if you walk me through it it would help.

File: backends/tfhe-cuda-backend/cuda/include/linear_algebra.h [hunk 1]
```diff
@@ -27,6 +27,12 @@ void cuda_add_lwe_ciphertext_vector_64(void *stream, uint32_t gpu_index,
                                        void const *lwe_array_in_2,
                                        uint32_t input_lwe_dimension,
                                        uint32_t input_lwe_ciphertext_count);
+void cuda_add_lwe_ciphertext_vector_64_with_packing(
```
[the remark this answers is not in the document]
Comment (guillermo-oyarzun) [reply backends/tfhe-cuda-backend/cuda/include/linear_algebra.h:30]: it is something needed for the signed overflowing add/sub, not actually tested in this PR, I could remove it and just included in the other PR we will have

File: tfhe/src/integer/gpu/server_key/radix/add.rs [hunk 2]
```diff
@@ -227,6 +285,18 @@ impl CudaServerKey {
         streams.synchronize();
     }
 
+    pub fn unchecked_add_assign_with_packing<T: CudaIntegerRadixCiphertext>(
```
[the remark this answers is not in the document]
Comment (guillermo-oyarzun) [reply tfhe/src/integer/gpu/server_key/radix/add.rs:288]: it is something needed for the signed overflowing add/sub, not actually tested in this PR, I could remove it and just included in the other PR we will have

File: backends/tfhe-cuda-backend/cuda/include/integer/integer_utilities.h [hunk 3]
```diff
@@ -917,204 +1028,1465 @@ template <typename Torus> struct int_overflowing_sub_memory {
...
+      auto carry_flag_last = lut_carry_flag_last->get_lut(gpu_indexes[0], 0);
+
+      generate_device_accumulator<Torus>(
+          streams[0], gpu_indexes[0], carry_flag_last, glwe_dimension,
+          polynomial_size, message_modulus, carry_modulus, f_carry_last);
+
+      lut_carry_flag_last->broadcast_lut(streams, gpu_indexes, gpu_indexes[0]);
+    }
+
+    active_gpu_count = get_active_gpu_count(2 * num_radix_blocks, gpu_count);
```
[the remark this answers is not in the document]
Comment (guillermo-oyarzun) [reply backends/tfhe-cuda-backend/cuda/include/integer/integer_utilities.h]: not really i copied it wrongly from other part.

File: backends/tfhe-cuda-backend/cuda/include/integer/integer.h [hunk 4]
```diff
@@ -297,6 +297,47 @@ void cleanup_cuda_propagate_single_carry(void *const *streams,
...
+    void *lwe_array, void *carry_out, const void *carry_in, int8_t *mem_ptr,
+    void *const *bsks, void *const *ksks, uint32_t num_blocks,
+    uint32_t requested_flag, uint32_t uses_carry);
+
+void cleanup_cuda_fast_propagate_single_carry(void *const *streams,
+                                              uint32_t const *gpu_indexes,
+                                              uint32_t gpu_count,
+                                              int8_t **mem_ptr_void);
+
+void scratch_cuda_integer_overflowing_sub_kb_64_inplace(
```
[the remark this answers is not in the document]
Comment (guillermo-oyarzun) [reply backends/tfhe-cuda-backend/cuda/include/integer/integer.h:320]: I removed some unnecessary headers that were still there

File: backends/tfhe-cuda-backend/cuda/include/integer/integer_utilities.h [hunk 5]
```diff
@@ -917,204 +1028,1465 @@ template <typename Torus> struct int_overflowing_sub_memory {
...
+          streams, gpu_indexes, gpu_count, params, num_groups,
+          big_lwe_size_bytes, true);
+    }
+
+    free(h_scalar_array_cum_sum);
+    free(h_second_lut_indexes);
+  };
+
+  // needed for the division to update the lut indexes
+  void update_lut_indexes(cudaStream_t const *streams,
```
[the remark this answers is not in the document]
Comment (guillermo-oyarzun) [reply backends/tfhe-cuda-backend/cuda/include/integer/integer_utilities.h:1794]: I moved most of the logic to the division now

File: backends/tfhe-cuda-backend/cuda/include/linear_algebra.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (guillermo-oyarzun) [reply backends/tfhe-cuda-backend/cuda/include/linear_algebra.h:30]: change it to make it look like the cpu

**Review by agnesLeroy — APPROVED** [review]

**Pull request merged**